editor/src/messages/tool/tool_messages/select_tool.rs

```diff
@@ -435,32 +553,164 @@ impl Fsm for SelectToolFsmState {
 						RotatingBounds
 					} else if intersection.last().map(|last| selected.contains(last)).unwrap_or(false) {
 						responses.push_back(DocumentMessage::StartTransaction.into());
-
 						tool_data.layers_dragging = selected;
-
 						tool_data
 							.snap_manager
 							.start_snap(document, input, document.bounding_boxes(Some(&tool_data.layers_dragging), None, render_data), true, true);
 
 						Dragging
 					} else {
-						if !input.keyboard.get(add_to_selection as usize) {
+						responses.push_back(DocumentMessage::StartTransaction.into());
+						if !input.keyboard.get(add_to_selection as usize) && tool_data.selected_type == LayerSelectionBehavior::Deepest {
 							responses.push_back(DocumentMessage::DeselectAllLayers.into());
 							tool_data.layers_dragging.clear();
 						}
-
 						if let Some(intersection) = intersection.pop() {
 							tool_data.layer_selected_on_start = Some(intersection.clone());
-							selected = vec![intersection];
-							responses.push_back(DocumentMessage::AddSelectedLayers { additional_layers: selected.clone() }.into());
-							responses.push_back(DocumentMessage::StartTransaction.into());
-							tool_data.layers_dragging.append(&mut selected);
-							tool_data
-								.snap_manager
-								.start_snap(document, input, document.bounding_boxes(Some(&tool_data.layers_dragging), None, render_data), true, true);
-
+							selected = vec![intersection.clone()];
+							// Shallowest manipulation
```

Comment:
Could we extract some of this to separate functions to reduce the size of this match block?